### ISSUE TITLE
YAML Configuration | Radiosondes

### ISSUE THREAD

**Issue by danholdaway** [GEOS-ESM/swell#78]
## Agreement with GSI using GeoVaLs in UFO (Hofx)

- [ ] Done

## Agreement with GSI using GeoVaLs in UFO (Quality control)

- [ ] Done

## Agreement with GSI using Swell

- [ ] Done

## Current issues:

- [ ] https://github.com/JCSDA-internal/ufo/pull/2218

**Comment by gmao-msienkie** [comment]
Original yaml file fails to be processed -  the input data is missing 'LaunchTime' values.  If a station has more than one launch in a synoptic time, the launch time would be used to group the obs from each launch separately.  In the GSI the obs error inflation when multiple obs exist inside vertical model layers is done report-by-report, hence, sounding by sounding.  Without launch time the multiple launches from a station would be grouped together for that filter.

**Comment by danholdaway** [comment]
![sondes_gsihofxbc_vs_hofx_air_temperature](https://user-images.githubusercontent.com/27729500/190192023-ff96587e-89b5-4568-9275-edac52d2097a.png)
![sondes_gsihofxbc_vs_hofx_eastward_wind](https://user-images.githubusercontent.com/27729500/190192025-43d654b8-d47e-4125-be62-165cf4fb4194.png)
![sondes_gsihofxbc_vs_hofx_northward_wind](https://user-images.githubusercontent.com/27729500/190192028-690d5558-5896-4034-8d40-37f04b3e38c4.png)
![sondes_gsihofxbc_vs_hofx_specific_humidity](https://user-images.githubusercontent.com/27729500/190192029-47392bfc-2d28-4f93-a599-34e1bca94a45.png)
![sondes_gsihofxbc_vs_hofx_surface_pressure](https://user-images.githubusercontent.com/27729500/190192032-ae0c9d7a-93b6-4ed8-ae27-472eaba36641.png)
![sondes_gsihofxbc_vs_hofx_virtual_temperature](https://user-images.githubusercontent.com/27729500/190192036-fb10680b-c461-4a24-8b76-2f851fb5091a.png)

**Comment by gmao-jjin3** [comment]
![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/199291521-baa848cd-fe6a-4138-aeaf-6dbb2a1d44e9.png)
Observational errors in JEDI and GSI. There is no duplication inflation in JEDI yet.  A question, how to re-set observational error range [100-300 Pa] after the *ObsErrorFactorSfcPressure@ObsFunction* function?   The following *bounds check* filters after it have no impact.  [cory.r.martin@noaa.gov](mailto:cory.r.martin@noaa.gov)
```
  # error inflation based on Ps correction
  - filter: Perform Action
    filter variables:
    - name: surface_pressure
    action:
      name: inflate error
      inflation variable:
        name: ObsErrorFactorSfcPressure@ObsFunction
        options:
          geovar_geomz: geopotential_height
          geovar_sfc_geomz: surface_geopotential_height
  # ObsError >= 100
  - filter: Bounds Check
    filter variables:
    - name: surface_pressure@ObsErrorData
    where:
    - variable: surface_pressure@ObsErrorData
      maxvalue: 100
    action:
      name: assign error
      error parameter: 100
  # ObsError <= 300
  - filter: Bounds Check
    filter variables:
    - name: surface_pressure@ObsErrorData
    where:
    - variable: surface_pressure@ObsErrorData
      minvalue: 300
    action:
      name: assign error
      error parameter: 300

```

**Comment by gmao-jjin3** [comment]
I tried another way to re-set the obs error range:
```
  # ObsError >= 100
  - filter: Variable Assignment
    where:
    - variable:
        #name: surface_pressure@ObsError
        name: ObsErrorData/surface_pressure
      maxvalue: 100
    assignments:
    #- name: surface_pressure@ObsError
    - name: ObsErrorData/surface_pressure
      value: 100
  # ObsError <= 300
  - filter: Variable Assignment
    where:
    - variable:
        #name: surface_pressure@ObsError
        name: ObsErrorData/surface_pressure
      minvalue: 300
    assignments:
    #- name: surface_pressure@ObsError
    - name: ObsErrorData/surface_pressure
      value: 300
```
However, JEDI this time thinks "ObsErrorData" is a new variable and asks an input of data  'type'. However, the UFO document says it is the updated obs error:
https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/jedi-components/ufo/qcfilters/introduction.html#observation-errors

Anyway, Here are steps to handle a surface pressure error in GSI:
1, Assign an initial error (from the error table)
2, ObsErrorFactorSfcPressure@ObsFunction
3, Reset ObsError between Min ~ Max
4, Gross error check:  (O-F)/ObsError
5, Inflate the error, which is the value before Step 3, because of duplication, and make it the final obs error.

**Comment by gmao-jjin3** [comment]
![corscat_omb_vs_gsiombbc](https://user-images.githubusercontent.com/61251050/199586019-35b5279b-d7c5-4f1f-9194-4d9e4d9a2871.png)
These O-B in UFO are close to those in GSI except for a few spots, likely because of some missing QC.

**Comment by gmao-jjin3** [comment]
Here is an issue to re-set ObsError within a range after initial inflation.
https://github.com/JCSDA-internal/ufo/issues/2363

**Comment by gmao-jjin3** [comment]
![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/202229063-19bbb34c-506f-49fa-8136-31176c4a7914.png)
Now the observational errors can be re-bounded correctly after initial inflation. See Yaml settings in https://github.com/JCSDA-internal/ufo/issues/2363.

**Comment by gmao-jjin3** [comment]
![corscat_qc_omb_vs_gsiombbc](https://user-images.githubusercontent.com/61251050/203086463-633d9640-1469-41fc-8b70-7f71194d4f11.png)
Now UFO can reproduce GSI surface pressure Hofx after a bug fix in surface pressure correction. Sondes' "height", not "station_elevation" should be used in the correction. See the UFO pull-request: https://github.com/JCSDA-internal/ufo/pull/2493

**Comment by danholdaway** [comment]
Great stuff @gmao-jjin3! Thanks! When making these scatter plots can you make them for all observations rather than what passed QC? There should be a match for all obs regardless of whether each system used the observation or not.

**Comment by gmao-jjin3** [comment]
@danholdaway  Now I can see that the combining programs within the original IODA-converter package messed up observations. Here are surface_pressure and virtual_temperature in those separated and combined files.

```
For_others/conv_obs_check% python ck_ps.py
/discover/nobackup/drholdaw/JediData/GMAOReferenceRun/geos_to_r2d2/observations/conventional/obs_sep/sondes_ps_obs_2020121500.nc4
--Obs with z=3650m:
i,type,subtype,sta_id,lat,lon,z(3650m),ps: 203 120 0 b'55591   ' 29.67 91.13 3650.0 65110.0
For_others/conv_obs_check% python ck_tv.py
/discover/nobackup/drholdaw/JediData/GMAOReferenceRun/geos_to_r2d2/observations/conventional/obs_sep/sondes_tv_obs_2020121500.nc4
--Obs with z=3650m:
i,type,subtype,sta_id,lat,lon,z(3650m),tv: 6570 120 0 b'55591   ' 29.67 91.13 3650.0 269.85
For_others/conv_obs_check% python ck_tv_combined.py
/discover/nobackup/drholdaw/JediData/GMAOReferenceRun/geos_to_r2d2/observations/conventional/obs/sondes_obs_.nc4
--Obs with z=3650m:
i,type,subtype,sta_id,lat,lon,z(3650m),tv,ps: 28708 -2147483647 -2147483647 b'55591   ' 29.67 91.13 3650.0 **9.96921e+36 65110.0**
i,type,subtype,sta_id,lat,lon,z(3650m),tv,ps: 28710 120 0 b'55591   ' 29.67 91.13 3650.0 **269.85 9.96921e+36**

```

**Comment by danholdaway** [comment]
Is this what is leading to the difference between GSI and UFO? Good to know if you've found the cause. Do you have a sense for what is going wrong in the combine tool? This is coming from ioda converters.

**Comment by gmao-jjin3** [comment]
It is a reason. Not sure what is wrong in the combine Python programs at the moment.

**Comment by gmao-jjin3** [comment]
Okay.  There is a very tiny difference in "air_pressure" in separated conventional IODA files.  This values is saved in hPa in GSI' surface_pressure output and is multiplied by 100 in the IODA converter, while it is saved in Pa in outputs for surface temperature. They are supposed to be same in converted IODA files.  However, the combine Python program doesn't think they are same because of the tiny difference!  As a result, the combine program messed up temperature and pressure at this location.      

```
For_others/conv_obs_check% python  ck_tv.py 
/discover/nobackup/drholdaw/JediData/GMAOReferenceRun/geos_to_r2d2/observations/conventional/obs_sep/sondes_tv_obs_2020121500.nc4
--Obs with z=3650m:
air_pressure 65110.0
i,type,subtype,sta_id,lat,lon,z(3650m),tv: 6570 120 0 b'55591   ' 29.67 91.13 3650.0 269.85
For_others/conv_obs_check% python ck_ps.py
/discover/nobackup/drholdaw/JediData/GMAOReferenceRun/geos_to_r2d2/observations/conventional/obs_sep/sondes_ps_obs_2020121500.nc4
--Obs with z=3650m:
air_pressure 65109.996
i,type,subtype,sta_id,lat,lon,z(3650m),ps: 203 120 0 b'55591   ' 29.67 91.13 3650.0 65110.0
For_others/conv_obs_check% python ck_tv_combined.py 
/discover/nobackup/drholdaw/JediData/GMAOReferenceRun/geos_to_r2d2/observations/conventional/obs/sondes_obs_.nc4
--Obs with z=3650m:
air_pressure 65109.996
i,type,subtype,sta_id,lat,lon,z(3650m),tv,ps: 28708 -2147483647 -2147483647 b'55591   ' 29.67 91.13 3650.0 9.96921e+36 65110.0
air_pressure 65110.0
i,type,subtype,sta_id,lat,lon,z(3650m),tv,ps: 28710 120 0 b'55591   ' 29.67 91.13 3650.0 269.85 9.96921e+36

```

**Comment by danholdaway** [comment]
OK. So the combine program needs to be fitted with some tolerance for deciding if the obs are the same then. Or perhaps using some smarter logic to decide if the obs are the same. Tagging @CoryMartin-NOAA. I believe EMC worked on this combine program so may have a quicker idea of how to modify it.

**Comment by CoryMartin-NOAA** [comment]
@danholdaway I think this will be difficult to do because it is using (if I recall correctly) a python method that uses exact values to determine if it is the same location. I'm not sure how easy this will be to implement.

**Comment by danholdaway** [comment]
OK, thanks for letting us know. I don't think we'll be able to replicate the GSI without the combine step since there are filters requiring multiple variables so we'll have to think on it. For example I think temperature is used on one of the surface pressure filters for example. Thanks for the work to establish what's happening Jianjun. We'll have to think a bit on the correct solution.

**Comment by CoryMartin-NOAA** [comment]
We can probably only check lat/lon/time and ignore pressure?

**Comment by gmao-jjin3** [comment]
Noop, we cannot ignore pressure because sonde data are profiles. It may work for surface observations but we very likely get into troubles for other observations. I am re-making those IODA data after changing GSI. However, it is hard to guarantee there is no machine errors like these and it still is a risk not to have any tolerance in co-locating data even my new data work.

**Comment by gmao-jjin3** [comment]
![corscat_gsihofxbc_vs_hofx](https://user-images.githubusercontent.com/61251050/205819237-a422d55b-2c62-467a-b9e7-469c00debbbf.png)
This figure shows JEDI HofX are identical to GSI Hofx before QC.  It looks like that there is no issue with my new data. However, we don't really solve the combine issue.

**Comment by gmao-jjin3** [comment]
![corscat_eferr_x_GsiDup_vs_finaler](https://user-images.githubusercontent.com/61251050/208170473-a574536a-df86-49ee-81c0-0d3aabcd43f4.png)

This figure shows UFO surface pressure errors timed by those GSI duplicate factors are equal to GSI final observation errors. I think this figure confirms that surface pressure errors are set up correctly before we finalize the duplicate factor function.

**Comment by gmao-jjin3** [comment]
![corscat_EffectiveQC_vs_GsiEffectiveQC](https://user-images.githubusercontent.com/61251050/208171211-f270ee26-2c34-4529-90db-d1eaad4bce8d.png)

@danholdaway  This figure shows GSI effective QC flags vs UFO effective QC flags for all of the data.  It puzzles me because that I find those GSI and UFO flags have the same number of zero values in the output file /discover/nobackup/jjin3/jedi/JediUfoTests/Output/sondes_qc_2020121500_0000.nc4, which is not what this figure shows.  This figure is made with configuration between lines 341-365 in /discover/nobackup/jjin3/jedi/JediUfoTests/Eva/Config/sondes.yaml.  My short Python program to check those GSI and UFO QC flags is /discover/nobackup/jjin3/jedi/For_others/conv_obs_check/ioda_sondes_ps/ck_ps_qc.py . Can you do me a favor to take a look at the yaml file and the python program and see why they show different results? Thanks.

**Comment by gmao-jjin3** [comment]
@danholdaway Never mind.  I think I need remove undefined values during plotting.

**Comment by gmao-jjin3** [comment]
![corscat_EffectiveQC_vs_GsiEffectiveQC](https://user-images.githubusercontent.com/61251050/208183503-dbe89125-6f33-4cd2-bca5-220a4114d489.png)
This figure suggests that there are same numbers zero values in GSI effective QC flags and UFO effective QC flags.

**Comment by gmao-jjin3** [comment]
![density_qc_omb_vs_gsiombbc](https://user-images.githubusercontent.com/61251050/208478481-8107628c-e722-4bad-b030-9ce2bd02a3be.png)

Histograms of surface pressure O-F after QC. They are identical in UFO and GSI.

**Comment by gmao-jjin3** [comment]
![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/210626452-69e42587-36d2-4943-be62-4aee70c525c1.png)

The duplicate factor does work well for surface_pressure at the moment.

**Comment by gmao-jjin3** [comment]
![corscat_eferr_vs_finaler check_missing](https://user-images.githubusercontent.com/61251050/211057012-42aa2c77-85ac-469d-b31e-afdada1e4dee.png)

This is the latest comparison between UFO and GSI final surface_pressure observational errors.  Obviously, they do not agree to each other. The reason is that the duplicate_factor filter only checks locations of sondes data within current UFO framework which processes sondes and other surface pressure observations separately. GSI, however, looks for co-located  surface pressure observations within all observation data types.  Here is a debug in the "setupps.f90"
```
              tfact=min(one,abs(data(itime,k)-data(itime,l))/dfact1)
              dup(k)=dup(k)+one-tfact*tfact*(one-dfact)
              dup(l)=dup(l)+one-tfact*tfact*(one-dfact)
              if(ictype(nint(data(ikxx,k))) == 120 ) then
                write(6,"(A,i4,i9,f9.4,a,2f9.3,A,i4,i9,f9.4,a,2f9.3)")  &
                           "JJJ, type(k),k,dup(k)", ictype(nint(data(ikxx,k))),k,dup(k), &
                           ", lat,lon,", data(ilate,k),data(ilone,k), &
                           "; type(l),l, dup(l)", ictype(nint(data(ikxx,l))),l, dup(l), &
                           ", lat,lon,", data(ilate,l),data(ilone,l)
              endif
```
   and you can see **dulplicate_factors are calculated for colocated  surface (kx/datatype=181 or 187) and sondes data (kx/datatype=120):** from the ouput (only a few lines are posted here)
```
   JJJ, type(k),k,dup(k) 120        1   2.0000, lat,lon,   42.870  129.500; type(l),l, dup(l) 120        6   2.0000, lat,lon,   42.8
70  129.500
JJJ, type(k),k,dup(k) 120        1   2.8594, lat,lon,   42.870  129.500; type(l),l, dup(l) 181      129   1.8594, lat,lon,   42.8
70  129.500
JJJ, type(k),k,dup(k) 120        1   1.9375, lat,lon,   43.050  141.330; type(l),l, dup(l) 181       28   1.9375, lat,lon,   43.0
50  141.330
JJJ, type(k),k,dup(k) 120        1   2.8750, lat,lon,   43.050  141.330; type(l),l, dup(l) 181       47   1.9375, lat,lon,   43.0
50  141.330
JJJ, type(k),k,dup(k) 120        1   3.7014, lat,lon,   43.050  141.330; type(l),l, dup(l) 181       59   1.8264, lat,lon,   43.0
50  141.330
JJJ, type(k),k,dup(k) 120        1   1.9992, lat,lon,   62.530   17.450; type(l),l, dup(l) 187        2   1.9992, lat,lon,   62.5
30   17.450
JJJ, type(k),k,dup(k) 120        1   2.9961, lat,lon,   62.530   17.450; type(l),l, dup(l) 187        3   1.9969, lat,lon,   62.5
30   17.450
JJJ, type(k),k,dup(k) 120        1   3.7731, lat,lon,   62.530   17.450; type(l),l, dup(l) 187        4   1.7770, lat,lon,   62.5
30   17.450
JJJ, type(k),k,dup(k) 120        1   4.5231, lat,lon,   62.530   17.450; type(l),l, dup(l) 187        5   1.7500, lat,lon,   62.5
30   17.450
JJJ, type(k),k,dup(k) 120        1   5.3927, lat,lon,   62.530   17.450; type(l),l, dup(l) 187      271   1.8696, lat,lon,   62.5
30   17.450
```

In short, we have to reconsider how to process those surface_pressure and other conventional observations in UFO. @danholdaway @CoryMartin-NOAA. 
BTW, there is an update in checking missing values of PreUseFlgs atop of Akira's source code:
https://github.com/JCSDA-internal/ufo/commit/92f6206864e9cda1731fbd14e5c435da7b4b3432

**Comment by CoryMartin-NOAA** [comment]
I'm not sure this is something we are going to be able to do in JEDI given the `obs space` approach... Perhaps @danholdaway has a different view/suggestion but I'm inclined to think this is going to be nearly impossible given how IODA is constructed.

**Comment by gmao-jjin3** [comment]
@CoryMartin-NOAA I agree. I don't see a solution for this.

**Comment by gmao-msienkie** [comment]
I was looking at some output that I had produced for the sondes running the ObsErrorFactorConventional filter and came across a sounding from station 50953 that looks like it was duplicated.  The obs are slightly different on some levels though.  In any case the two soundings were processed separately through errormod, but were treated as all one sounding in UFO by ObsErrorFactorConventional because they had the same launch time (time of first sounding level was the same for both, though they had different DHR sounding times).  In any case, that's another reason for the output from UFO not matching GSI.  
[SID=50953_duplicate_sounding.txt](https://github.com/GEOS-ESM/swell/files/10576313/SID.50953_duplicate_sounding.txt)

**Comment by gmao-msienkie** [comment]
I made a list of the ADPUPA data in the gdas.prepbufr file from 20201215/00z and found there were quite a few (87!) duplicated soundings on that date.  (From the station IDs they appear to be Chinese sondes.) Possibly some of the soundings were transmitted both in BUFR and text format.  In any case, the GSI 'errormod' filter would treat the soundings separately while the UFO filter would filter the soundings together if the reported release time were the same.   While the DHR values given for the sounding pairs were different, the drift time for the first level of the soundings was the same.  Actually the drift lat,lon,time was the same for quite a few levels (if not most) and so would trigger down-weighting by the duplicate check.
[dup_sonde_adpupa.txt](https://github.com/GEOS-ESM/swell/files/10580469/dup_sonde_adpupa.txt)

**Comment by gmao-jjin3** [comment]
![density_qc_omb_vs_gsiombbc b](https://user-images.githubusercontent.com/61251050/217442307-1b257be1-4f20-4e34-8fa1-164d1489d19d.png)
![density_qc_omb_vs_gsiombbc](https://user-images.githubusercontent.com/61251050/217442316-6263bdc3-49d1-466e-b1c8-6d5f93267604.png)

![corscat_eferr_x_GsiDup_vs_finaler](https://user-images.githubusercontent.com/61251050/217442428-4d1a2fc2-0988-4388-8c03-444d57f97a1f.png)

![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/218523086-dae409fb-5281-4218-8e56-91a978727f61.png)



UFO O-F and observational errors for **airTemperature** match GSI values when observational errors made by UFO's function "ObsErrorFactorConventional" are replaced by GSI adjusted errors are used.

**Comment by gmao-jjin3** [comment]
![density_qc_omb_vs_gsiombbc b](https://user-images.githubusercontent.com/61251050/217443345-e923d394-9802-4552-b76e-684365c0e008.png)
![density_qc_omb_vs_gsiombbc](https://user-images.githubusercontent.com/61251050/217443348-e1a2c13f-756f-48b5-814e-84d887bc60af.png)
![corscat_eferr_x_GsiDup_vs_finaler](https://user-images.githubusercontent.com/61251050/217443390-1e314272-63e6-436f-9645-9bd9b783c08d.png)

UFO O-F and observational errors for **virtualTemperature** do not match those values in GSI.

**Comment by gmao-jjin3** [comment]
Jianjun,

Myself and Hui Shao reviewed the GSI code quite a lot, yes.
Using error inflation as a means to deal with too many data points in a single grid volume is very questionable as a technique instead of alternate means, but, yes, we're aware of the GSI code for doing error inflation.  We are trying our best to match GSI, but ultimately, that is not our goal.  Rather, we desire to make the best DA use possible.  We strive to make a first attempt to match GSI.  However, this is not always guaranteed as sufficient time to produce an identical match is difficult.

Please feel free to open a GitHub/Zenhub issue if things are not optimal.

Greg


On Tue, Feb 7, 2023 at 2:03 PM Jin, Jianjun (GSFC-610.1)[SCIENCE SYSTEMS AND APPLICATIONS INC] <[jianjun.jin@nasa.gov](mailto:jianjun.jin@nasa.gov)> wrote:
Hi Cory and Greg
 
Are you aware of any UFO codes to deal with temperature obs error inflation in GSI’s setupt.f90:
https://github.com/CoryMartin-NOAA/GSI/blob/ea04c69c4a3c29fc2109b6d7cdd21dfd049b1cf0/src/gsi/setupt.f90#L819

Thank you,
 
Jianjun

**Comment by gmao-jjin3** [comment]
![density_qc_omb_vs_gsiombbc b](https://user-images.githubusercontent.com/61251050/218513334-a62e9928-f873-4141-b349-b4287f8d0781.png)
![density_qc_omb_vs_gsiombbc](https://user-images.githubusercontent.com/61251050/218513338-34cba48f-c43b-4218-8674-89c66c56c3b5.png)
![corscat_eferr_x_GsiDup_vs_finaler](https://user-images.githubusercontent.com/61251050/218513360-56abe896-85c8-4bc0-ae70-9afcf0051994.png)


![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/218521239-c2b6b1de-e212-4622-a9ce-18555245a35f.png)


Now  UFO O-F and observational errors match GSI values using (1) Wei's pressure-check inflator, and (2) GSI adjust error in Yaml configuration.

**Comment by gmao-jjin3** [comment]
Many thanks to @gmao-wgu for writing the pressure check function. Here are a few updates in my directory: /discover/nobackup/jjin3/jedi/jedi-work2/Wei_Gu/ioda_v3_temperature

1) Variable names were updated as in ioda_v3.
2) Added the function for temperature observations as in setupt.f90 in GSI.
3) Removed "GSI adjust error" from the function. We use it as a temporary solution and we can configure it in YAML files, so that we can remove it easily in future. BTW, I also copied my configuration files into that directory.

**Comment by danholdaway** [comment]
Thanks for uploading these plots @gmao-jjin3. In each case can you also provide the plots without the multiplication by GsiDup? For posterity it would be good to just get a sense of how many observations this is impacting.

**Comment by gmao-jjin3** [comment]
@danholdaway  Added a figure showing comparisons between UFO final error without GSI duplicator versus GSI final observation error (see above).  There are not much difference between them in sonde observations, but there are a lot differences in ship observations.

**Comment by asewnath** [comment]
<img width="759" alt="Screen Shot 2023-02-15 at 3 34 04 PM" src="https://user-images.githubusercontent.com/14252652/219148458-353c41d2-ae61-40f4-b6f1-1e3c0d7c9070.png">

With duplicate check filter added, some duplicates not accounted for
<img width="688" alt="Screen Shot 2023-02-15 at 3 34 30 PM" src="https://user-images.githubusercontent.com/14252652/219148609-4823856a-bcbb-4574-bfb5-aff700657eb9.png">

**Comment by gmao-jjin3** [comment]
Duplicate factor is applied. 

![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/219155312-c0dc61f6-c124-4d47-9f65-c2ef8637e480.png)
![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/219155524-34ed1599-2c31-4010-9cf3-8da492562b90.png)

**Comment by gmao-wgu** [comment]
**The consistency check for Radiosonde wind (kx=220, 221) is completed, which can be demonstrated from the following plots**:

**The number of observations passing QC in the GSI**:

![sondes_density_qced_omb_vs_gsiomb_gsi_stats](https://user-images.githubusercontent.com/59488002/220687909-311b7162-51bf-4f38-a99f-5cf7f5800759.png)

**The number of observations passing QC in Jedi**:

![sondes_density_qced_omb_vs_gsiomb_jedi_stats](https://user-images.githubusercontent.com/59488002/220688119-f5f5a9fd-27bb-448e-9762-c5356798ab7e.png)

**The final observation error comparison**:

![sondes_corscat_qced_eferr_vs_gsifinal](https://user-images.githubusercontent.com/59488002/220688429-0257ebd1-2d9a-4855-a111-18b0e5801d58.png)

**Comment by gmao-msienkie** [comment]
> I made a list of the ADPUPA data in the gdas.prepbufr file from 20201215/00z and found there were quite a few (87!) duplicated soundings on that date. (From the station IDs they appear to be Chinese sondes.) Possibly some of the soundings were transmitted both in BUFR and text format. In any case, the GSI 'errormod' filter would treat the soundings separately while the UFO filter would filter the soundings together if the reported release time were the same. While the DHR values given for the sounding pairs were different, the drift time for the first level of the soundings was the same. Actually the drift lat,lon,time was the same for quite a few levels (if not most) and so would trigger down-weighting by the duplicate check.

I was looking at this issue again because it came up when I was checking how the wind obs were handled by ObsErrorFactorConventional versus the GSIAdjustObsError.  It seems to me that the agreement would be improved if we changed the definition of 'time_launch' in the GSI for raobs so that it uses the DHR (like the aircraft) rather than the first time level of the drfdat.  It looks like the pairs of soundings have different DHR values (for whatever reason) while they often have the same values for the initial drfdat time.   That puts the two reports in the same sounding for ObsErrorFactorConventional while the GSI processes the sounding reports in 'errormod' separately.   It would be better to process those together (or actually better to screen for duplicates better in the preprocessing) but for the purpose of a match with GSI I think the change in the definition would help.

**Comment by gmao-jjin3** [comment]
Thank you @gmao-msienkie .  I made the change in read_prepbufr.f90. Let's see if we can get better results tomorrow.

**Comment by gmao-jjin3** [comment]
![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/231540062-0ba8795e-6458-4583-8c90-59182bfeb547.png)

The  final observation errors do match after duplication is considered only for the same observation type such as sondes, surface, etc. in GSI.  See the previous comparison when duplicate factor is considered for surface pressure in all conventional data: https://github.com/GEOS-ESM/swell/issues/78#issuecomment-1373909089

**Comment by gmao-jjin3** [comment]
![corscat_eferr_vs_finaler](https://user-images.githubusercontent.com/61251050/231542176-51240d3e-370f-4bae-a97d-2219a30cbee4.png)

@gmao-msienkie 's suggestion about launch_time does make a difference, removed one of the out of line two spots in the last figure posted on Feb 15.  https://github.com/GEOS-ESM/swell/issues/78#issuecomment-1432016186

**Comment by gmao-jjin3** [comment]
Just add links of PR to "Replace station_elevation by height for Ps correction in GSI":
https://github.com/JCSDA-internal/ufo/pull/2493
(There are some discussions about extrapolation: https://github.com/JCSDA-internal/ufo/pull/2493#issuecomment-1329760749) 
https://github.com/JCSDA-internal/ufo/pull/2594
https://github.com/JCSDA-internal/ufo/pull/2609

**Comment by gmao-jjin3** [comment]
![density_qc_omb_vs_gsiomb](https://github.com/GEOS-ESM/swell/assets/61251050/cbc773ff-572c-45d0-ad4b-0cf326576f21)
![corscat_qc_omb_vs_hofx-GsiHofXBc](https://github.com/GEOS-ESM/swell/assets/61251050/07439a4d-4bdb-46c7-92d5-f1e72620a949)
![corscat_eferr_x_GsiDup_vs_finaler](https://github.com/GEOS-ESM/swell/assets/61251050/d7826028-e113-4b24-882f-f0b6163686ff)
![corscat_eferr_vs_finalerr](https://github.com/GEOS-ESM/swell/assets/61251050/c2ded5c1-cf73-489e-b136-610959d91e4a)
Sonde surface pressures from x0048.  Final observational error cannot be reproduced by UFO.

**Comment by gmao-jjin3** [comment]
Just add a note that surface pressure hofx  values in UFO match  those in GSI after a bug fix which is merged in UFO developed branch: https://github.com/JCSDA-internal/ufo/pull/2987
![corscat_qc_omb_vs_hofx-GsiHofXBc](https://github.com/GEOS-ESM/swell/assets/61251050/88ed437f-b083-420a-97cc-42f7968c1cd0)

**Comment by gmao-jjin3** [comment]
Validation against GEOS x0048 results (/discover/nobackup/jjin3/jedi/For_others/x0048.jj_20230818/conv/combine_conv/obs/sondes_obs_2021121200.nc4)

![density_qc_omb_vs_gsiomb](https://github.com/GEOS-ESM/swell/assets/61251050/57fd3775-51c8-41bb-a5e9-9124851e4cc5)
![corscat_qc_omb_vs_hofx-GsiHofXBc](https://github.com/GEOS-ESM/swell/assets/61251050/50aabe78-7bdf-4edb-9618-f4dc69eb2f77)
![corscat_EffectiveQC_vs_GsiEffectiveQC](https://github.com/GEOS-ESM/swell/assets/61251050/95ab750d-94cc-4dc4-be84-049bb33e6498)
![corscat_eferr_x_GsiDup_vs_finaler](https://github.com/GEOS-ESM/swell/assets/61251050/0b20c5ad-a349-4064-b123-291a3c0f64a8)
![corscat_eferr_vs_finalerr](https://github.com/GEOS-ESM/swell/assets/61251050/ce8ec663-4b30-405c-8d71-eeac6516e0d6)

**Comment by gmao-jjin3** [comment]
Validation against GEOS x0048 results (/discover/nobackup/jjin3/jedi/For_others/x0048.jj_20230818/conv/combine_conv/obs/sondes_obs_2021121200.nc4)
![density_qc_omb_vs_gsiomb](https://github.com/GEOS-ESM/swell/assets/61251050/c4c94f75-8d2e-493d-abde-9c10b0a8a2cf)
![corscat_qc_omb_vs_hofx-GsiHofXBc](https://github.com/GEOS-ESM/swell/assets/61251050/86d810dd-8f3a-41ac-b33c-875ce13e40d3)
![corscat_EffectiveQC_vs_GsiEffectiveQC](https://github.com/GEOS-ESM/swell/assets/61251050/b5d452e2-17ac-4c60-ae60-6b440d8fad9c)
![corscat_eferr_x_GsiDup_vs_finaler](https://github.com/GEOS-ESM/swell/assets/61251050/435e90bb-0c4f-4154-96bd-d1fd7a0d592f)
![corscat_eferr_vs_finalerr](https://github.com/GEOS-ESM/swell/assets/61251050/641cace4-e483-4f08-8c6f-44dd3b8eabad)
